File: examples_cpp/src/controllers/pid_trajectory_geometric_controller.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file pid_trajectory_geometric_controller.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "controllers/pid_trajectory_geometric_controller.hpp"

#include <yaml-cpp/yaml.h>

#include <chrono>
#include <cstddef>
#include <stdexcept>
#include <utility>

#include "utils/example_config_utils.hpp"

namespace mpc_examples::adapters {

namespace {

pid_controller::PIDParameters<double> parsePidParameters(const YAML::Node& node,
                                                         const std::string& section) {
  pid_controller::PIDParameters<double> params;
  const auto read_vec3 = [&](const char* key) -> Eigen::Vector3d {
    const YAML::Node child = node[key];
    if (!child) {
      throw std::invalid_argument(section + "." + key + " is required.");
    }
    return detail::readVector<3>(child, section + "." + key);
  };

  params.Kp_gains = read_vec3("kp");
  params.Ki_gains = read_vec3("ki");
  params.Kd_gains = read_vec3("kd");

  if (node["antiwindup_cte"]) {
    const double v =
        detail::readDoubleRequired(node["antiwindup_cte"], section + ".antiwindup_cte");
    params.antiwindup_cte = Eigen::Vector3d::Constant(v);
  }
  if (node["alpha"]) {
    const double v = detail::readDoubleRequired(node["alpha"], section + ".alpha");
    params.alpha   = Eigen::Vector3d::Constant(v);
  }
  if (node["a_max"]) {
    const double v = detail::readDoubleRequired(node["a_max"], section + ".a_max");
    if (v > 0.0) {
      params.upper_output_saturation = Eigen::Vector3d::Constant(v);
      params.lower_output_saturation = Eigen::Vector3d::Constant(-v);
    }
  }
  return params;
}

}  // namespace

PidTrajectoryGeometricController::PidTrajectoryGeometricController(const Config& cfg)
    : cfg_(cfg) {
  if (cfg_.v_max <= 0.0) {
    throw std::invalid_argument("PidTrajectoryGeometricController: v_max must be > 0.");
  }
}

PidTrajectoryGeometricController::Config PidTrajectoryGeometricController::loadConfigFromYaml(
    const std::string& path) {
  const YAML::Node root = detail::loadYamlRoot(path);
  Config cfg;

  if (root["v_max"]) {
    cfg.v_max = detail::readDoubleRequired(root["v_max"], "v_max");
  }

  const YAML::Node ctrl = root["controller"];
  if (!ctrl || !ctrl.IsMap()) {
    throw std::invalid_argument("pid_trajectory_geometric config: 'controller' must be a mapping.");
  }
  if (!ctrl["trajectory"]) {
    throw std::invalid_argument(
        "pid_trajectory_geometric config: 'controller.trajectory' is required.");
  }
  cfg.trajectory_pid_params = parsePidParameters(ctrl["trajectory"], "controller.trajectory");

  if (!ctrl["geometric"] || !ctrl["geometric"].IsMap()) {
    throw std::invalid_argument(
        "pid_trajectory_geometric config: 'controller.geometric' must be a mapping.");
  }
  const YAML::Node geo = ctrl["geometric"];
  if (!geo["mass"]) {
    throw std::invalid_argument(
        "pid_trajectory_geometric config: 'controller.geometric.mass' is required.");
  }
  cfg.attitude_params.vehicle_mass =
      detail::readDoubleRequired(geo["mass"], "controller.geometric.mass");
  if (!geo["rotation_kp"]) {
    throw std::invalid_argument(
        "pid_trajectory_geometric config: 'controller.geometric.rotation_kp' is required.");
  }
  cfg.rates_params.kp_rotation =
      detail::readVector<3>(geo["rotation_kp"], "controller.geometric.rotation_kp");

  return cfg;
}

void PidTrajectoryGeometricController::initialize(const mav_model::State& /*initial_state*/,
                                                  const ExampleConfig& example_cfg) {
  control_period_ = example_cfg.pid_dt;
  if (control_period_ <= 0.0) {
    throw std::invalid_argument("PidTrajectoryGeometricController: example_cfg.pid_dt must be > 0.");
  }

  pid_controllers::TrajectoryControllerParameters<double> traj_params;
  traj_params.pid_parameters = cfg_.trajectory_pid_params;
  traj_ctrl_ = std::make_unique<pid_controllers::TrajectoryController<double>>(traj_params);

  geo_ctrl_ = std::make_unique<geometric_controller::GeometricController<double>>(
      cfg_.attitude_params, cfg_.rates_params);
}

framework::ControlCommand PidTrajectoryGeometricController::computeCommand(
    const mav_model::State& state,
    const std::vector<framework::ReferenceSample>& references) {
  if (references.empty()) {
    throw std::invalid_argument("PidTrajectoryGeometricController: references must not be empty.");
  }
  const framework::ReferenceSample& ref = references.front();

  const Eigen::Vector3d position       = state.getPositionVector();
  const Eigen::Vector3d velocity       = state.getLinearVelocityVector();
  const Eigen::Quaterniond orientation = state.getOrientationVector();

  const auto t0 = std::chrono::high_resolution_clock::now();

  const Eigen::Vector3d acc_des = traj_ctrl_->trajectoryToLinearAcceleration(
      position, velocity, ref.position, ref.velocity, Eigen::Vector3d::Zero(), control_period_);

  const auto [thrust, rates] = geo_ctrl_->accelerationToRates(acc_des, ref.yaw, orientation);

  const auto t1  = std::chrono::high_resolution_clock::now();
  last_solve_us_ = std::chrono::duration<double>(t1 - t0).count() * 1e6;

  framework::ControlCommand cmd;
  cmd.thrust_n     = thrust;
  cmd.angular_rate = rates;
  return cmd;
}

}  // namespace mpc_examples::adapters

File: examples_py/examples_py/controllers/pid_trajectory_geometric_controller.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Direct trajectory PID + geometric attitude controller adapter.

Python mirror of
``examples_cpp/src/controllers/pid_trajectory_geometric_controller.cpp``.
Wraps :class:`mavpy.controllers.pid_controllers.TrajectoryController` and
:class:`mavpy.controllers.geometric_controller.GeometricController`
behind :class:`IController` for trajectory-based references.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import os
import time
from dataclasses import dataclass, field
from typing import List, Optional

import numpy as np
import yaml
from mavpy.controllers.geometric_controller import (
    AttitudeGeometricControllerParameters,
    GeometricController,
    GeometricControllerParameters,
    RatesGeometricControllerParameters,
)
from mavpy.controllers.libs.pid_controller import PIDParameters
from mavpy.controllers.pid_controllers import (
    TrajectoryController,
    TrajectoryControllerParameters,
)
from mavpy.model import State

from examples_py.framework import (
    ControlCommand,
    ExampleConfig,
    IController,
    ReferenceField,
    ReferenceSample,
    make_mask,
)


@dataclass
class PidTrajectoryGeometricConfig:
    """Parsed configuration for :class:`PidTrajectoryGeometricController`."""

    trajectory_pid_params: PIDParameters = field(default_factory=PIDParameters)
    attitude_params: AttitudeGeometricControllerParameters = field(
        default_factory=AttitudeGeometricControllerParameters)
    rates_params: RatesGeometricControllerParameters = field(
        default_factory=RatesGeometricControllerParameters)
    v_max: float = 1.0


def _read_vec3(node, name: str) -> np.ndarray:
    if not isinstance(node, (list, tuple)) or len(node) != 3:
        raise ValueError(f'{name} must be a sequence with 3 elements.')
    return np.array([float(x) for x in node], dtype=float)


def _parse_pid_params(node: dict, section: str) -> PIDParameters:
    params = PIDParameters()
    if 'kp' not in node:
        raise ValueError(f'{section}.kp is required.')
    if 'ki' not in node:
        raise ValueError(f'{section}.ki is required.')
    if 'kd' not in node:
        raise ValueError(f'{section}.kd is required.')
    params.Kp_gains = _read_vec3(node['kp'], f'{section}.kp')
    params.Ki_gains = _read_vec3(node['ki'], f'{section}.ki')
    params.Kd_gains = _read_vec3(node['kd'], f'{section}.kd')
    if 'antiwindup_cte' in node:
        value = float(node['antiwindup_cte'])
        params.antiwindup_cte = np.full(3, value, dtype=float)
    if 'alpha' in node:
        value = float(node['alpha'])
        params.alpha = np.full(3, value, dtype=float)
    if 'a_max' in node:
        value = float(node['a_max'])
        if value > 0.0:
            params.upper_output_saturation = np.full(3, value, dtype=float)
            params.lower_output_saturation = np.full(3, -value, dtype=float)
    return params


class PidTrajectoryGeometricController(IController):
    """Direct trajectory PID for smooth trajectory tracking (horizon size 1).

    Pipeline, executed once per control period:
      1. trajectory PID: (state.pos, state.vel, ref.pos, ref.vel) -> acc_des
         (parallel pos/vel feedback into a single PID; no acceleration feedforward)
      2. geometric:      (acc_des, ref.yaw, state.orientation) -> (thrust, rates)

    Requires ReferenceField.POSITION | VELOCITY.
    Suitable for smooth trajectory generators (jerk_limited, gcopter, dynamic,
    mav_traj_gen) and trajectory MPC.
    """

    def __init__(self, cfg: PidTrajectoryGeometricConfig) -> None:
        if cfg.v_max <= 0.0:
            raise ValueError('PidTrajectoryGeometricController: v_max must be > 0.')
        self._cfg = cfg
        self._traj_ctrl: Optional[TrajectoryController] = None
        self._geo_ctrl: Optional[GeometricController] = None
        self._control_period = 0.01
        self._last_solve_us = 0.0
        self._name = 'PidTrajectoryGeometricController'

    @staticmethod
    def load_config_from_yaml(path: str) -> PidTrajectoryGeometricConfig:
        """Load a :class:`PidTrajectoryGeometricConfig` from a YAML file.

        Expected structure (config_pid_trajectory.yaml)::

            v_max: 3.0
            controller:
              trajectory: {kp, ki, kd, antiwindup_cte, alpha}
              geometric:  {mass, rotation_kp}
        """
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            root = yaml.safe_load(f)
        if not isinstance(root, dict):
            raise ValueError('pid_trajectory_geometric config: root must be a mapping.')

        cfg = PidTrajectoryGeometricConfig()
        if 'v_max' in root:
            cfg.v_max = float(root['v_max'])

        ctrl = root.get('controller')
        if not isinstance(ctrl, dict):
            raise ValueError("pid_trajectory_geometric config: 'controller' must be a mapping.")
        if 'trajectory' not in ctrl:
            raise ValueError("pid_trajectory_geometric config: 'controller.trajectory' is required.")
        cfg.trajectory_pid_params = _parse_pid_params(ctrl['trajectory'], 'controller.trajectory')

        geo = ctrl.get('geometric')
        if not isinstance(geo, dict):
            raise ValueError("pid_trajectory_geometric config: 'controller.geometric' must be a mapping.")
        if 'mass' not in geo:
            raise ValueError("pid_trajectory_geometric config: 'controller.geometric.mass' is required.")
        if 'rotation_kp' not in geo:
            raise ValueError(
                "pid_trajectory_geometric config: 'controller.geometric.rotation_kp' is required.")
        cfg.attitude_params = AttitudeGeometricControllerParameters()
        cfg.attitude_params.vehicle_mass = float(geo['mass'])
        cfg.rates_params = RatesGeometricControllerParameters()
        cfg.rates_params.kp_rotation = _read_vec3(
            geo['rotation_kp'], 'controller.geometric.rotation_kp')
        return cfg

    def initialize(self, initial_state: State, example_cfg: ExampleConfig) -> None:
        self._control_period = example_cfg.pid_dt
        if self._control_period <= 0.0:
            raise ValueError('PidTrajectoryGeometricController: example_cfg.pid_dt must be > 0.')

        traj_params = TrajectoryControllerParameters()
        traj_params.pid_parameters = self._cfg.trajectory_pid_params
        self._traj_ctrl = TrajectoryController(traj_params)

        geo_params = GeometricControllerParameters(
            attitude_parameters=self._cfg.attitude_params,
            rates_parameters=self._cfg.rates_params,
        )
        self._geo_ctrl = GeometricController(geo_params)

    def reference_horizon_size(self) -> int:
        return 1

    def reference_horizon_dt(self) -> float:
        return self._control_period

    def control_period(self) -> float:
        return self._control_period

    def compute_command(
        self,
        state: State,
        references: List[ReferenceSample],
    ) -> ControlCommand:
        if not references:
            raise ValueError('PidTrajectoryGeometricController: references must not be empty.')
        assert self._traj_ctrl is not None
        assert self._geo_ctrl is not None
        ref = references[0]

        position = np.asarray(state.position, dtype=float)
        velocity = np.asarray(state.linear_velocity, dtype=float)
        orientation = np.asarray(state.orientation, dtype=float)

        t0 = time.perf_counter()
        acc_des = self._traj_ctrl.trajectory_to_linear_acceleration(
            position,
            velocity,
            np.asarray(ref.position, dtype=float),
            np.asarray(ref.velocity, dtype=float),
            np.zeros(3),
            self._control_period)

        thrust, rates = self._geo_ctrl.acceleration_to_rates(
            acc_des, float(ref.yaw), orientation)
        t1 = time.perf_counter()
        self._last_solve_us = (t1 - t0) * 1e6

        return ControlCommand(thrust_n=float(thrust),
                              angular_rate=np.asarray(rates, dtype=float))

    def required_reference_fields(self) -> ReferenceField:
        return make_mask([
            ReferenceField.POSITION,
            ReferenceField.VELOCITY,
        ])

    def name(self) -> str:
        return self._name

    def last_solve_time_micros(self) -> float:
        return self._last_solve_us
